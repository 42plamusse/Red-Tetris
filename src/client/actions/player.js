import socket from "../services/socket-api";
import {
  DISCONNECT_PLAYER,
  NEW_PLAYER,
  UPDATE_PLAYER,
  UPDATE_PLAYERS_LIST
} from "../../constants/constants";
import { updateRoom } from "./room";

export const connectPlayer = playerName => {
  return dispatch => {
    socket.emit(NEW_PLAYER, playerName, response => {
      if (response.status === "success") {
        // dispatch(updatePlayer(response.playerInfo));
      }
    });
  };
};

export const disconnectPlayer = () => {
  return dispatch => {
    socket.emit(DISCONNECT_PLAYER);
    dispatch(updateRoom({}));
    window.location.hash = "";
    dispatch(updatePlayer({}));
  };
};

export const updatePlayer = player => ({
  type: UPDATE_PLAYER,
  player
});

export const updatePlayersList = players => ({
  type: UPDATE_PLAYERS_LIST,
  players
});
