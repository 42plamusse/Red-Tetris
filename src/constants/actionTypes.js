export const JOIN_ROOM = "JOIN_ROOM";
export const LEAVE_ROOM = "LEAVE_ROOM";
export const UPDATE_ROOM = "UPDATE_ROOM";
export const UPDATE_ROOMS = "UPDATE_ROOMS";

export const NEW_PLAYER = "NEW_PLAYER";
export const UPDATE_PLAYER = "UPDATE_PLAYER";
export const DISCONNECT_PLAYER = "DISCONNECT_PLAYER";

export const CONNECTION = "connection";
export const DISCONNECT = "disconnect";

export const UPDATE_SPECTRES = "UPDATE_SPECTRES";

export const KEY_PRESSED = "KEY_PRESSED";

export const START_GAME = "START_GAME";
