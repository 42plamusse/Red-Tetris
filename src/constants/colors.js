export const MALUS = "#3d3d3d";
