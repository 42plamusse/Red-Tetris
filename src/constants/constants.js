export const MALUS = "#3d3d3d";

export const JOIN_ROOM = "JOIN_ROOM";

export const UPDATE_ROOM = "UPDATE_ROOM";

export const LEAVE_ROOM = "LEAVE_ROOM";

export const NEW_PLAYER = "NEW_PLAYER";

export const DISCONNECT_PLAYER = "DISCONNECT_PLAYER";

export const UPDATE_PLAYER = "UPDATE_PLAYER";

export const UPDATE_PLAYERS_LIST = "UPDATE_PLAYERS_LIST";

export const HANDLE_HASH = "HANDLE_HASH";

export const UPDATE_ROOMS = "UPDATE_ROOMS";

export const DELETE_PLAYER = "DELETE_PLAYER";

export const SHOW_TOAST = "SHOW_TOAST";

export const CONNECTION = "connection";

export const UPDATE_SPECTRES = "UPDATE_SPECTRES";

export const ADD_CHAT_MESSAGE = "ADD_CHAT_MESSAGE";
export const NEW_CHAT_MESSAGE = "NEW_CHAT_MESSAGE";

export const DISCONNECT = "disconnect";

export const KEY_PRESSED = "KEY_PRESSED";
export const START_GAME = "START_GAME";

export const enumKeys = {
  ARROW_LEFT: "ArrowLeft",
  ARROW_RIGHT: "ArrowRight",
  ARROW_DOWN: "ArrowDown",
  ARROW_UP: "ArrowUp",
  SPACE: "Space",
  ENTER: "Enter"
};
export const ARROW_LEFT = "ArrowLeft";
export const ARROW_RIGHT = "ArrowRight";
export const ARROW_DOWN = "ArrowDown";
export const ARROW_UP = "ArrowUp";
export const SPACE = "Space";
export const ENTER = "Enter";

export const UPDATE_BOARD = "UPDATE_BOARD";
export const MAX_PLAYER = 5;

export const LOBBY_ROOM = "LOBBY_ROOM";

export const LOG_LINE =
  "\n_________________________________________________________________________________________________________\n\n";
