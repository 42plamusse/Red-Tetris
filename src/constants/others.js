export const MAX_PLAYER = 10;

export const LOBBY_ROOM = "LOBBY_ROOM";

export const LOG_LINE =
  "\n_________________________________________________________________________________________________________\n\n";
