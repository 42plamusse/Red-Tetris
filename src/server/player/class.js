class Player {
  constructor(name = "", id = "") {
    this._name = name;
    this._id = id;
    this._room = null;
    this._board = Array(200).fill(0);
  }
  get name() {
    return this._name;
  }

  get id() {
    return this._id;
  }
  get board() {
    return this._board;
  }

  set board(board) {
    this._board = board;
  }

  setBoardAtIndex(index, value) {
    this._board[index] = value;
  }

  get room() {
    return this._room;
  }

  set room(room) {
    this._room = room;
  }

  createPublicPlayerObject() {
    const player = {};
    player.name = this._name;
    player.id = this._id;
    player.board = this._board;
    return player;
  }
}

export default Player;
