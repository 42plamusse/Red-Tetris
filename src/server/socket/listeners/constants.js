export const CREATE_ROOM = "CREATE_ROOM";

export const NEW_PLAYER = "NEW_PLAYER";

export const FETCH_ROOMS = "FETCH_ROOMS";

export const HASH_CHANGED = "HASH_CHANGED";

export const NEW_ROOM_LIST = "NEW_ROOM_LIST";
