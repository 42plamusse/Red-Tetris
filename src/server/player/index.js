import Game from "../game/class";
import Player from "../player/class";
import { leaveRoom } from "../room";
import { isAlphaNumeric } from "../../utils/utils";
import { LOBBY_ROOM } from "../../constants/constants";

export const connectPlayer = (playerName, callback, socket) => {
  console.log("[CALL] connectPlayer");
  if (!isAlphaNumeric(playerName) || playerName.length > 12) {
    callback({
      status: "error",
      message: "Player name must be 1 to 12 alphanumeric characters long"
    });
  } else {
    const player = Game.findPlayer(socket.id);
    if (!player) {
      const player = Game.addPlayer(new Player(playerName, socket.id));
      const playerInfo = player.createPublicPlayerObject();
      callback({ status: "success", playerInfo });
      console.log("[UPDATED] after connectPlayer", Game);
      return player;
    }
  }
};

export const disconnectPlayer = (socket, io) => {
  console.log("[CALL] disconnectPlayer on : ", socket.id);
  leaveRoom(socket, io);
  Game.removePlayer(socket.id);
  socket.leave(LOBBY_ROOM);
  console.log("[UPDATED] after disconnectPlayer", Game);
};
