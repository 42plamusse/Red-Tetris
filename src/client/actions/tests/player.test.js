import thunk from "redux-thunk";
import configureMockStore from "redux-mock-store";
import { connectPlayer, disconnectPlayer, updatePlayer } from "../player";
import { UPDATE_PLAYER } from "../../../constants/actionTypes";
import socket from "../../services/socket-api";

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

beforeEach(done => {
  // Setup
  if (socket.disconnected) {
    socket.connect();
  }
  socket.on("connect", () => {
    done();
  });
});

afterEach(done => {
  //   Cleanup;
  if (socket.connected) {
    socket.disconnect();
  }
  done();
});

describe("player actions", () => {
  test("disconnectPlayer clean room, player and hash", done => {
    const store = mockStore({
      player: { name: "Bob", id: "ID", isHost: true, board: [0, 0, 0] }
    });
    window.location.hash = "#test[bob]";
    store.dispatch(disconnectPlayer());
    setTimeout(() => {
      store.getActions().forEach(action => expect(action).toMatchSnapshot());
      expect(window.location.hash).toBe("");
      done();
    }, 50);
  });
  test("connectPlayer updates player state with a valid name", done => {
    const store = mockStore({ player: {} });
    const playerName = "Bob";
    store.dispatch(connectPlayer(playerName));
    setTimeout(() => {
      const action = store.getActions()[0];
      delete action.player.id;
      expect(action).toMatchSnapshot();
      done();
    }, 50);
  });
  test("connectPlayer does not update player state with an unvalid name", done => {
    const store = mockStore({ player: {} });
    store.dispatch(connectPlayer("Bobadkjfhaskjdfghajkhsf"));
    setTimeout(() => {
      expect(store.getActions()[0]).toMatchSnapshot();
      done();
    }, 50);
  });
  test("updatePlayer create the corresponding action", () => {
    const player = { name: "Bob", id: "ID", isHost: true, board: [0, 0, 0] };
    const expectedAction = {
      type: UPDATE_PLAYER,
      player
    };
    expect(updatePlayer(player)).toEqual(expectedAction);
  });
});
