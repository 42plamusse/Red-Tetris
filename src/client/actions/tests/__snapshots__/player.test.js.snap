// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`player actions connectPlayer does not update player state with an unvalid name 1`] = `undefined`;

exports[`player actions connectPlayer updates player state with a valid name 1`] = `
Object {
  "player": Object {
    "board": Array [
      Array [
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
      ],
      Array [
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
      ],
      Array [
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
      ],
      Array [
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
      ],
      Array [
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
      ],
      Array [
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
      ],
      Array [
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
      ],
      Array [
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
      ],
      Array [
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
      ],
      Array [
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
      ],
      Array [
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
      ],
      Array [
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
      ],
      Array [
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
      ],
      Array [
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
      ],
      Array [
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
      ],
      Array [
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
      ],
      Array [
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
      ],
      Array [
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
      ],
      Array [
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
      ],
      Array [
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
      ],
    ],
    "isHost": null,
    "name": "Bob",
  },
  "type": "UPDATE_PLAYER",
}
`;

exports[`player actions disconnectPlayer clean room, player and hash 1`] = `
Object {
  "room": Object {},
  "type": "UPDATE_ROOM",
}
`;

exports[`player actions disconnectPlayer clean room, player and hash 2`] = `
Object {
  "player": Object {},
  "type": "UPDATE_PLAYER",
}
`;
